System/File/OsPath.hs

```diff
@@ -93,8 +118,18 @@ appendFile' fp contents = withFile fp AppendMode (`BS.hPut` contents)
 
 -- | Open a file and return the 'Handle'.
 openFile :: OsPath -> IOMode -> IO Handle
-openFile (OsString fp) = P.openFile fp
+openFile osfp@(OsString fp) = augmentError "openFile" osfp . P.openFile fp
 
 -- | Open an existing file and return the 'Handle'.
 openExistingFile :: OsPath -> IOMode -> IO Handle
-openExistingFile (OsString fp) = P.openExistingFile fp
+openExistingFile osfp@(OsString fp) = augmentError "openExistingFile" osfp . P.openExistingFile fp
+
+addFilePathToIOError :: String -> OsPath -> IOException -> IOException
+addFilePathToIOError fun fp ioe = unsafePerformIO $ do
+  fp'  <- either (const (fmap OSP.toChar . OSP.unpack $ fp)) id <$> try @SomeException (OSP.decodeFS fp)
```

Comment:
This matches the logic in unix: https://github.com/haskell/unix/blob/7db23ecad7593210ce38c48a462be6c50d080e00/System/Posix/PosixPath/FilePath.hsc#L161

Except we're not in IO and so I use `SomeException` and try to force the result as best as possible.